// @rcdoc/cli@0.1.29/dist/common/constant.js
import {
  BABEL_CONFIG_FILE,
  CJS_DIR,
  CLI_DIR,
  CWD,
  JEST_CONFIG_FILE,
  MDOC_BUILD_CONFIG_FILE,
  PACKAGE_ENTRY_FILE,
  PACKAGE_JSON_FILE,
  PACKAGE_STYLE_FILE,
  PROJECT_CACHE_DIR,
  PROJECT_CJS_DIR,
  PROJECT_CLI_DIST_DIR,
  PROJECT_DIST_DIR,
  PROJECT_DOCS_DIR,
  PROJECT_ES_DIR,
  PROJECT_HD_DIR,
  PROJECT_POSTCSS_CONFIG_FILE,
  PROJECT_SITE_DIST_DIR,
  PROJECT_SRC_DIR,
  PROJECT_STYLE_DIR,
  ROOT,
  SCRIPT_EXTS,
  SITE_CUSTOM_COMPONENT_FILE,
  SITE_SHARD_CONFIG_FILE,
  SITE_SHARED_API_FILE,
  SITE_SHARED_FILE,
  SITE_SHARED_MENU_FILE,
  SITE_SHARED_NAV_FILE,
  SITE_SHARED_ROUTES_FILE,
  SITE_SRC_DIR,
  STYLE_DEPS_JSON_FILE,
  STYLE_EXTS,
  getPackageJson
} from "../chunk-D3YCLLHO.js";
import "../chunk-OKPPX6YV.js";
export {
  BABEL_CONFIG_FILE,
  CJS_DIR,
  CLI_DIR,
  CWD,
  JEST_CONFIG_FILE,
  MDOC_BUILD_CONFIG_FILE,
  PACKAGE_ENTRY_FILE,
  PACKAGE_JSON_FILE,
  PACKAGE_STYLE_FILE,
  PROJECT_CACHE_DIR,
  PROJECT_CJS_DIR,
  PROJECT_CLI_DIST_DIR,
  PROJECT_DIST_DIR,
  PROJECT_DOCS_DIR,
  PROJECT_ES_DIR,
  PROJECT_HD_DIR,
  PROJECT_POSTCSS_CONFIG_FILE,
  PROJECT_SITE_DIST_DIR,
  PROJECT_SRC_DIR,
  PROJECT_STYLE_DIR,
  ROOT,
  SCRIPT_EXTS,
  SITE_CUSTOM_COMPONENT_FILE,
  SITE_SHARD_CONFIG_FILE,
  SITE_SHARED_API_FILE,
  SITE_SHARED_FILE,
  SITE_SHARED_MENU_FILE,
  SITE_SHARED_NAV_FILE,
  SITE_SHARED_ROUTES_FILE,
  SITE_SRC_DIR,
  STYLE_DEPS_JSON_FILE,
  STYLE_EXTS,
  getPackageJson
};
