// @rcdoc/cli@0.1.29/dist/chunk-3JZJACPY.js
import {
  genSiteMenu
} from "./chunk-YI6P2V7Z.js";
import {
  genSiteRoutesShared
} from "./chunk-2PZ2BHXQ.js";
import {
  genSiteMenuShared
} from "./chunk-WYVUPVZU.js";
import {
  getCssLang
} from "./chunk-67CKMM2A.js";
import {
  normalizePath,
  replaceExt,
  smartOutputFile
} from "./chunk-PLIM7P7F.js";
import {
  context_default
} from "./chunk-WOQRSCMF.js";
import {
  PACKAGE_STYLE_FILE,
  SITE_SHARD_CONFIG_FILE,
  SITE_SHARED_FILE,
  SITE_SHARED_MENU_FILE,
  getPackageJson
} from "./chunk-D3YCLLHO.js";

// src/compiler/gen-site-shared.ts
import slash from "slash2";
function genImportJSON() {
  return `
export { default as config} from '${normalizePath(SITE_SHARD_CONFIG_FILE)}';
export { default as menus} from '${normalizePath(SITE_SHARED_MENU_FILE)}';
  `;
}
function genStyles() {
  const CSS_LANG = getCssLang();
  return `import '${slash(replaceExt(PACKAGE_STYLE_FILE, `.${CSS_LANG}`))}'`;
}
function genExportVersion() {
  const projectPackageJson = getPackageJson();
  return `export const packageVersion = '${projectPackageJson.version}';`;
}
function genSiteDesktopShared() {
  const { menus, routes } = genSiteMenu();
  const code = `${genImportJSON()}
${context_default.opts.site.injectComponentCss ? genStyles() : ""}
${genExportVersion()}
`;
  genSiteMenuShared(menus);
  genSiteRoutesShared(routes);
  smartOutputFile(SITE_SHARED_FILE, code);
}

export {
  genSiteDesktopShared
};
