// @rcdoc/cli@0.1.29/dist/chunk-D3YCLLHO.js
// src/common/constant.ts
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
var __dirname = dirname(fileURLToPath(import.meta.url));
var CWD = process.cwd();
var ROOT = CWD;
var PROJECT_DIST_DIR = join(ROOT, "dist");
var PROJECT_ES_DIR = join(PROJECT_DIST_DIR, "es");
var PROJECT_CJS_DIR = join(PROJECT_DIST_DIR, "cjs");
var PROJECT_HD_DIR = join(PROJECT_DIST_DIR, "2x");
var PROJECT_DOCS_DIR = join(ROOT, "docs");
var PACKAGE_JSON_FILE = join(ROOT, "package.json");
var PROJECT_POSTCSS_CONFIG_FILE = join(ROOT, "postcss.config.js");
var PROJECT_CACHE_DIR = join(ROOT, "node_modules/.cache");
var PROJECT_CLI_DIST_DIR = join(ROOT, ".rcdoc");
var PROJECT_SITE_DIST_DIR = join(ROOT, "docs-dist");
var SITE_SRC_DIR = join(__dirname, "..", "docs");
var PACKAGE_ENTRY_FILE = join(PROJECT_CLI_DIST_DIR, "package-entry.js");
var PACKAGE_STYLE_FILE = join(PROJECT_CLI_DIST_DIR, "package-style.css");
var SITE_SHARED_FILE = join(PROJECT_CLI_DIST_DIR, "site-shared.js");
var SITE_SHARED_MENU_FILE = join(PROJECT_CLI_DIST_DIR, "site-shared-menu.js");
var SITE_SHARED_NAV_FILE = join(PROJECT_CLI_DIST_DIR, "site-shared-nav.json");
var SITE_SHARED_API_FILE = join(PROJECT_CLI_DIST_DIR, "site-shared-api.json");
var SITE_SHARED_ROUTES_FILE = join(PROJECT_CLI_DIST_DIR, "site-shared-routes.js");
var SITE_CUSTOM_COMPONENT_FILE = join(PROJECT_CLI_DIST_DIR, "site-custom-component.js");
var SITE_SHARD_CONFIG_FILE = join(PROJECT_CLI_DIST_DIR, "config.json");
var STYLE_DEPS_JSON_FILE = join(PROJECT_CLI_DIST_DIR, "style-deps.json");
var CLI_DIR = join(__dirname, "..");
var CJS_DIR = join(__dirname, "..", "cjs");
var JEST_CONFIG_FILE = join(CJS_DIR, "jest.config.cjs");
var BABEL_CONFIG_FILE = join(CJS_DIR, "babel.config.cjs");
var MDOC_BUILD_CONFIG_FILE = join(CJS_DIR, ".rcdoc.build.cjs");
var SCRIPT_EXTS = [".js", ".jsx", ".ts", ".tsx"];
var STYLE_EXTS = [".css", ".less"];
function getPackageJson() {
  const rawJson = readFileSync(PACKAGE_JSON_FILE, "utf-8");
  return JSON.parse(rawJson);
}
var PROJECT_SRC_DIR = join(ROOT, "src");
var PROJECT_STYLE_DIR = join(PROJECT_SRC_DIR, "styles");

export {
  CWD,
  ROOT,
  PROJECT_DIST_DIR,
  PROJECT_ES_DIR,
  PROJECT_CJS_DIR,
  PROJECT_HD_DIR,
  PROJECT_DOCS_DIR,
  PACKAGE_JSON_FILE,
  PROJECT_POSTCSS_CONFIG_FILE,
  PROJECT_CACHE_DIR,
  PROJECT_CLI_DIST_DIR,
  PROJECT_SITE_DIST_DIR,
  SITE_SRC_DIR,
  PACKAGE_ENTRY_FILE,
  PACKAGE_STYLE_FILE,
  SITE_SHARED_FILE,
  SITE_SHARED_MENU_FILE,
  SITE_SHARED_NAV_FILE,
  SITE_SHARED_API_FILE,
  SITE_SHARED_ROUTES_FILE,
  SITE_CUSTOM_COMPONENT_FILE,
  SITE_SHARD_CONFIG_FILE,
  STYLE_DEPS_JSON_FILE,
  CLI_DIR,
  CJS_DIR,
  JEST_CONFIG_FILE,
  BABEL_CONFIG_FILE,
  MDOC_BUILD_CONFIG_FILE,
  SCRIPT_EXTS,
  STYLE_EXTS,
  getPackageJson,
  PROJECT_SRC_DIR,
  PROJECT_STYLE_DIR
};
