// @rcdoc/cli@0.1.29/dist/chunk-BNHXAF4L.js
import {
  ENTRY_EXTS,
  getExistFile,
  smartOutputFile
} from "./chunk-PLIM7P7F.js";
import {
  PROJECT_SRC_DIR,
  SITE_CUSTOM_COMPONENT_FILE
} from "./chunk-D3YCLLHO.js";

// src/compiler/gen-custom-component.ts
import path from "path";
import slash from "slash2";
function generateGuessPath(componentName) {
  return ENTRY_EXTS.map((ext) => `${componentName}.${ext}`);
}
var CUSTOM_COMPONENTS_ARR = [
  {
    path: ".",
    fileName: "Footer"
  },
  {
    path: ".",
    fileName: "HeaderExtra"
  },
  {
    fileName: "Header",
    exportName: "MobileHeader",
    exportDefault: "false",
    path: "./mobile"
  }
];
function genCustomComponent(item) {
  const componentPath = getExistFile({
    files: generateGuessPath(item.fileName).map((el) => path.join("_rcdoc_", item.path, el))
  });
  const exportName = item.exportName || item.fileName;
  const exportDefault = item.exportDefault || "() => null";
  if (componentPath) {
    return `export { default as ${exportName} } from '${slash(path.relative(PROJECT_SRC_DIR, componentPath))}'
`;
  }
  return `export const ${exportName} = ${exportDefault}
`;
}
var gen_custom_component_default = () => {
  const code = CUSTOM_COMPONENTS_ARR.reduce((a, v) => {
    a += genCustomComponent(v);
    return a;
  }, "");
  smartOutputFile(SITE_CUSTOM_COMPONENT_FILE, code);
};

export {
  genCustomComponent,
  gen_custom_component_default
};
